### DISCUSSION TITLE
How to use finn's load_external_weights() function?

### DISCUSSION THREAD

**Discussion by CHFang731** [Xilinx/finn#1099] [Q&A]
Hi everyone.

I have a question about finn's runtime_weights.I am currently trying to make partial reconfigurable for the cnn accelerator generated by finn. My block design is as shown in the figure (where StreamingDataflowPartition_1 is my PR part)

![image](https://github.com/Xilinx/finn/assets/43226625/76d969ec-0495-4641-b262-e5a09e007945)
![image](https://github.com/Xilinx/finn/assets/43226625/217b5da3-415b-400a-a91d-80277d79a6cd)

When I burned the full_bitstream, everything functioned normally and the model accuracy was in line with expectations.
However, when I burned the partial_bitstream, the model output results did not meet expectations, so I suspected that it was necessary to add the parameters of the model.
And I found that finn provided a function called `load_external_weights()`, in which the file format he required was a `.npy` file, so I found weight.npy in the MatrixVectorActivation folder generated by finn, as shown in the figure
![image](https://github.com/Xilinx/finn/assets/43226625/26782eb3-faae-44a4-83b1-40c39cfd5868)
![image](https://github.com/Xilinx/finn/assets/43226625/459d2d78-0ece-435b-a59c-dbb9afac8ed0)

and threw it into the corresponding runtime_weights folder, however, this does not seem to have any effect, the parameters do not seem to be loaded correctly, causing the final output of the model to still not be as expected.
And my question is
1. Are the parameters I am using correct, or do I need to generate the parameters again through finn or vivado?
2. finn also provides another function called load_runtime_weights(). What is the difference between it and load_external__weights()?

Thanks.

[finn_cnn.zip](https://github.com/user-attachments/files/15749144/finn_cnn.zip)

**Comment by fpjentzsch** [comment]
Hi,

load_external_weights() is for external weights. These are stored in DRAM and streamed into the FINN accelerator by a DMA core (similar to the DMAs for primary input/output generated by FINN in a full "ZynqBuild").

load_runtime_weights() is for runtime-configurable weights. These are stored in a FINN memstreamer IP with an AXI-lite interface inside the FPGA and streamed to the MVAU from there. The memory resource used for the weights can be set to LUTRAM, BRAM, or URAM. For LUTRAM and BRAM it is optional to (re-)load the weights as they are contained within the bitstream. For URAM you must load them as the URAM content cannot be set with the bitstream.

**Comment by DaZai3219** [comment]
hi, were you able to utilize "external" memory mode?
